Add parallel processing for CPU-intensive operations

bobtista opened the pull request:
Introduces thread pool implementation and parallelizes key CPU-intensive operations:
- Function disassembly in disassemble_function_matches
- Comparison record building in build_comparison_records
- Match bundle building in build_match_bundles
- Source line building in build_function_source_lines

Key changes:
- Add ThreadPool class for managing worker threads and task queues
- Add mutex protection for shared resources during parallel operations
- Add comprehensive thread safety documentation
- Keep I/O operations sequential for now
- Keep WorkQueue focused on command sequencing

Thread safety:
- Protected access to shared resources with mutex
- Independent processing where possible
- Clear documentation of thread-safe operations
- Exception handling in parallel tasks

bobtista committed:
Add parallel processing for CPU-intensive operations

Introduces thread pool implementation and parallelizes key CPU-intensive operations:
- Function disassembly in disassemble_function_matches
- Comparison record building in build_comparison_records
- Match bundle building in build_match_bundles
- Source line building in build_function_source_lines

Key changes:
- Add ThreadPool class for managing worker threads and task queues
- Add mutex protection for shared resources during parallel operations
- Add comprehensive thread safety documentation
- Keep I/O operations sequential for now
- Keep WorkQueue focused on command sequencing

Thread safety:
- Protected access to shared resources with mutex
- Independent processing where possible
- Clear documentation of thread-safe operations
- Exception handling in parallel tasks

bobtista committed:
Add error handling for parallel operations

Adds structured error handling for parallel processing:
- Add ErrorInfo struct to track errors during parallel execution
- Implement error collection with thread-safe mutex protection
- Add error reporting for failed function comparisons
- Keep errors contained without interrupting other parallel tasks

xezon commented on src/runner.cpp:
We should hold off with multithreading in Runner class until all functions for UI work commands are laid out.

But for the WorkQueue we can already explore and test the ThreadPool.

xezon commented on src/runner.cpp:
```diff
@@ -449,21 +464,58 @@ void Runner::build_match_bundle(
     const FunctionSetup setup1(*executable_pair[1], format);
 
+    // Create thread pool for parallel processing
+    ThreadPool pool;
```
Pool thread should perhaps account for UI thread and work queue thread. That always needs to run as well. So for example pool count may be `std::thread::hardware_concurrency - 2`.

Also, pool should probably be a singleton somewhere, otherwise there could be competing ThreadPools from different calls.

bobtista replied:
Check out threadpool_singelton.h

xezon commented on src/runner.cpp:
```diff
@@ -472,34 +524,77 @@ void Runner::build_function_source_lines(
+            {
+                match.comparison = AsmMatcher::run_comparison(match.function_pair, lookahead_limit);
+            }
+            catch (const std::exception &e)
```
We do not use exceptions, unless some 3rd party lib throws them (like cxxopts).

xezon commented on src/runner.cpp:
```diff
@@ -472,34 +524,77 @@ void Runner::build_function_source_lines(
+                    auto it = function_name_to_index_map.find(function_name);
+                    if (it != function_name_to_index_map.end())
+                    {
+                        std::lock_guard<std::mutex> lock(m_mutex); // Protect matches access
```
I do not like this approach. These locks will content and put threads to sleep.

Is this lock even necessary? There should only ever be one function to one set of source lines, no? Could be tested with atomic counter to see if it is accessed more than once.

xezon commented on src/runner.cpp:
```diff
@@ -472,34 +524,77 @@ void Runner::build_function_source_lines(
-
-                StringToIndexMapT::const_iterator it = function_name_to_index_map.find(function_name);
-                if (it != function_name_to_index_map.end())
+            results.push_back(pool.enqueue([&]() {
```
How expensive is this pushing and enqueuing in relation to the actual task? I think this needs to be measured in Profiler. It is possible that the overhead exceeds the cost of the tasks, minus the advantage from multi-threading.

For big batches, I would expect to give the each task a unique range to work on, and then return range results, and merge those range results from the future - in order - without using any mutex.

This accounts to all use cases.

xezon commented on src/threadpool.h:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    template<class F, class... Args>
+    auto enqueue(F &&f, Args &&...args) -> std::future<typename std::invoke_result<F, Args...>::type>
```
Can the return type be written without auto?

xezon commented on src/runner.cpp:
```diff
@@ -472,34 +524,77 @@ void Runner::build_function_source_lines(
 {
-    for (MatchedFunction &match : matches)
+    ThreadPool pool;
+    std::vector<std::future<void>> results;
```
Can the future only be void or can it also return some data with this thread pool?

xezon commented on src/threadpool.h:
```diff
@@ -0,0 +1,90 @@
+        m_condition.notify_all();
+        for (std::thread &worker : m_workers)
+        {
+            worker.join();
```
I think stop and join should be decoupled into separate functions for finer control, but still be called in destructor.

xezon commented on src/threadpool.h:
```diff
@@ -0,0 +1,90 @@
+            std::unique_lock<std::mutex> lock(m_queue_mutex);
+            if (m_stop)
+            {
+                throw std::runtime_error("enqueue on stopped ThreadPool");
```
Please no exceptions. Make enqueue return a bool instead.

xezon commented on src/threadpool.h:
```diff
@@ -0,0 +1,90 @@
+namespace unassemblize
+{
+
+class ThreadPool
```
Instead of using AI generated Thread Pool, perhaps pull a high performance open source ThreadPool from github.

xezon commented on src/workqueue.cpp:
```diff
@@ -140,8 +140,7 @@ void WorkQueue::ThreadRun()
         assert(command != nullptr);
         assert(command->has_work());
 
-        // #TODO: Spawn jobs for command to unclog the queue.
```
This comment is still relevant. The queue essentially needs to push command work to a global thread pool.

Note that there needs to be care if trying to use a global ThreadPool from within a thread of the global ThreadPool (is currently not the case), but also we cannot just have chains of large ThreadPools. So we need a suitable strategy here.

Perhaps for starters can have one small ThreadPool for the WorkQueue and a larger ThreadPool for the Runner.

xezon commented on src/threadpool.h:
```diff
@@ -0,0 +1,90 @@
+    {
+        for (size_t i = 0; i < num_threads; ++i)
+        {
+            m_workers.emplace_back([this] {
```
This could be more efficient by resizing `m_workers` before filling.

bobtista committed:
revert changes in runner

bobtista committed:
Add parallel command execution to WorkQueue

Changes:
- Add ThreadPool to WorkQueue for parallel command execution
- Move command execution to thread pool using lambda capture
- Keep command sequencing and callback handling intact

bobtista committed:
use BS thread pool, replacing the custom generated one

bobtista committed:
use singleton for threadpool, account for UI thread and work queue thread

bobtista commented on src/workqueue.cpp:
```diff
@@ -140,33 +142,40 @@ void WorkQueue::ThreadRun()
+        if (result == nullptr)
         {
-            result = std::make_unique<WorkQueueResult>();
+            const bool has_callback = command->has_callback();
```
Moved these consts here because command hasn't moved yet in this case. command->has_callback vs result->command->has_callback, I think it's safer this way

bobtista commented on src/workqueue.cpp:
```diff
@@ -140,33 +142,40 @@ void WorkQueue::ThreadRun()
             if (has_callback || has_delayed_command)
             {
-                m_callbackQueue.enqueue(std::move(result));
+                m_callbackQueue.enqueue(result);
```
std::move already happens here implicitly

xezon commented on src/workqueue.cpp:
```diff
@@ -128,9 +128,11 @@ void WorkQueue::ThreadRun()
+            {
                 // Queue is finally empty. Quit.
                 // Wait for jobs here when applicable.
                 break;
```
This class needs to wait for the busy thread pool to finish.
```
if (!m_commandQueue.try_dequeue(command))
{
  // Queue is finally empty. Quit.
  threadpool.join();
  break;
}
```

Or
```
if (!m_commandQueue.try_dequeue(command) && !threadpool.is_busy())
{
  // Queue is finally empty. Quit.
  break;
}
```

bobtista replied:
Resolved using futures `if (!m_commandQueue.try_dequeue(command) && futures.empty())`

xezon commented on src/workqueue.cpp:
```diff
@@ -140,33 +142,40 @@ void WorkQueue::ThreadRun()
+                                        .submit_task([cmd = std::move(command)]() { return cmd->work(); })
+                                        .get();
 
         m_lastFinishedCommandId = command->command_id;
```
Is using `command` after `move`.

bobtista replied:
Resolved - we now set the m_lastFinishedCommandId only after we find a completed task in the list of futures

xezon commented on src/workqueue.cpp:
```diff
@@ -140,33 +142,40 @@ void WorkQueue::ThreadRun()
-
-        WorkQueueResultPtr result = command->work();
+        // Submit work to thread pool
+        WorkQueueResultPtr result = ThreadPoolSingleton::get_workqueue_pool()
```
I assume `.get()` means it waits for the task to finish? If yes, then this effectively is single threaded, just with a fancy task invoked, no?

bobtista replied:
You're right - I've rewritten this to track and process pending tasks without blocking

bobtista committed:
Use futures to track and process command completion, processing tasks in parallel

xezon committed:
Merge branch 'generate-asm-ranges' into b/parallel

xezon commented:
This change does not compile:

```c++
9>D:\Projects\TheAssemblyArmada\unassemblize\3rdparty\readerwriterqueue\readerwriterqueue.h(562,20): error C2280: 'std::unique_ptr<unassemblize::WorkQueueResult,std::default_delete<unassemblize::WorkQueueResult>>::unique_ptr(const std::unique_ptr<unassemblize::WorkQueueResult,std::default_delete<unassemblize::WorkQueueResult>> &)': attempting to reference a deleted function
9>D:\Projects\TheAssemblyArmada\unassemblize\3rdparty\readerwriterqueue\readerwriterqueue.h(593,21): error C2280: 'std::unique_ptr<unassemblize::WorkQueueResult,std::default_delete<unassemblize::WorkQueueResult>>::unique_ptr(const std::unique_ptr<unassemblize::WorkQueueResult,std::default_delete<unassemblize::WorkQueueResult>> &)': attempting to reference a deleted function
9>D:\Projects\TheAssemblyArmada\unassemblize\3rdparty\readerwriterqueue\readerwriterqueue.h(614,27): error C2280: 'std::unique_ptr<unassemblize::WorkQueueResult,std::default_delete<unassemblize::WorkQueueResult>>::unique_ptr(const std::unique_ptr<unassemblize::WorkQueueResult,std::default_delete<unassemblize::WorkQueueResult>> &)': attempting to reference a deleted function
```

xezon commented:
This change has been obsoleted. Thread Pools have been implemented.
